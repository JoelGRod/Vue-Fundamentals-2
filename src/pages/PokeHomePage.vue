<template>
  <div v-if="!pokemon">
    <h1>Wait a moment please...</h1>
  </div>
  <div v-else>
    <h1>Recognize the Pokemon</h1>
    <PokemonImg :img-id="pokemon.id" :show-pokemon="showPokemon"> </PokemonImg>
    <AnswersList :pokemons="pokemons" @selection-pokemon="checkPokemon($event)">
    </AnswersList>
    <template v-if="showMessage">
      <h3>{{ message }}</h3>
    </template>
  </div>
</template>

<script>
// Components
import AnswersList from "@/components/AnswersList";
import PokemonImg from "@/components/PokemonImg";
// Helpers
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  name: "PokeHomePage",
  data() {
    return {
      pokemons: [],
      pokemon: null,
      showPokemon: false,
      message: "",
      showMessage: false,
    };
  },
  methods: {
    async pokemonsList() {
      this.pokemons = await getPokemonOptions()
      const rndInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemons[rndInt]
    },
    checkPokemon(pokemonId) {
      if (pokemonId === this.pokemon.id) {
        this.showPokemon = true
        this.message = "Well Done, you have a point!"
        this.showMessage = true
        this.nextPokemon()
      } else {
        this.message = "Wrong, try again..."
        this.showMessage = true
      }
    },
    nextPokemon() {
      setTimeout(() => {
        this.pokemon = null
        this.showMessage = false
        this.showPokemon = false
        this.pokemonsList()
      }, 3000)
    },
  },
  components: {
    AnswersList,
    PokemonImg,
  },
  // LifeCycle
  mounted() {
    this.pokemonsList()
  },
};
</script>

<style>
</style>