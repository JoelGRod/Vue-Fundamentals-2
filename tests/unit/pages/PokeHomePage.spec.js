import { shallowMount, mount } from '@vue/test-utils';
import PokeHomePage from '@/pages/PokeHomePage';
import { pokemonsList } from '../mocks/pokemons.mock';

describe('PokemonHomePage Component', () => {
    
    let wrapper;

    beforeEach( () => {
        wrapper = shallowMount( PokeHomePage );
    });

    test('should match with snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot();
    });

    test('should call pokemonsList when mounted', () => {
        // Arrange
        /** 
         * We need to spy over PokeHomePage 
         * method called in mounted BEFORE we mount the wrapper,
         * because this, we need to spy on literal object (passed 
         * by reference to wrapper) instead of wrapper
         */ 
        const pokemonsListSpy = jest.spyOn( PokeHomePage.methods, 'pokemonsList' );
        const wrapper = shallowMount(PokeHomePage);
        // Act
        // Assert
        expect(pokemonsListSpy).toHaveBeenCalledTimes(1);
    });

    test('should match snapshot when data is loaded', () => {
        // Arrange
        const wrapper = shallowMount( PokeHomePage, {
            data() {
                return {
                    pokemons: pokemonsList,
                    pokemon: pokemonsList[0],
                    showPokemon: false,
                    message: '',
                    showMessage: false
                }
            }
        })
        // Act
        // Assert
        expect( wrapper.html() ).toMatchSnapshot();
    });

    test('should show the PokemonImg and AnswersList components', () => {
        // Arrange
        const wrapper = shallowMount( PokeHomePage, {
            data() {
                return {
                    pokemons: pokemonsList,
                    pokemon: pokemonsList[0],
                    showPokemon: false,
                    message: '',
                    showMessage: false
                }
            }
        });

        // Because we used shallowMount
        const pokemonImg = wrapper.find('pokemon-img-stub');
        const answersList = wrapper.find('answers-list-stub');
        // Act
        // Assert
        expect(pokemonImg.exists()).toBeTruthy();
        expect(answersList.exists()).toBeTruthy();
        expect(pokemonImg.attributes('imgid')).toBe("1");
        expect(answersList.attributes('pokemons')).toBeTruthy();
    });

    // IMPORTANT!!
    test('test checkPokemon() function', async () => {
        // Arrange
        const wrapper = shallowMount( PokeHomePage, {
            data() {
                return {
                    pokemons: pokemonsList,
                    pokemon: pokemonsList[0],
                    showPokemon: false,
                    message: '',
                    showMessage: false
                }
            }
        });
        const nextPokemonSpy = jest.spyOn( wrapper.vm, 'nextPokemon' );
        // Act
        await wrapper.vm.checkPokemon(1);
        // Assert
        expect( wrapper.find('h3').exists() ).toBeTruthy();
        expect( wrapper.vm.showPokemon ).toBeTruthy();
        expect( wrapper.find('h3').text() ).toBe('Well Done, you have a point!');
        expect( nextPokemonSpy ).toBeCalledTimes(1);

        // Act
        await wrapper.vm.checkPokemon(2);
        // Assert
        expect( wrapper.vm.message ).toBe('Wrong, try again...');

    });

});