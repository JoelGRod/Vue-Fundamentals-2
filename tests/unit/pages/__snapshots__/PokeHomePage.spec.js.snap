// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonHomePage Component should match snapshot when data is loaded 1`] = `
<div>
  <h1>Recognize the Pokemon</h1>
  <pokemon-img-stub imgid="1" showpokemon="false"></pokemon-img-stub>
  <answers-list-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></answers-list-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonHomePage Component should match with snapshot 1`] = `
<div>
  <h1>Wait a moment please...</h1>
</div>
`;
